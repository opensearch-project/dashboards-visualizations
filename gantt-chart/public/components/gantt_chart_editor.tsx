/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiFieldNumber,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiPanel,
  EuiSpacer,
  EuiSuperSelect,
  EuiText,
  EuiTitle,
} from '@elastic/eui';
import React from 'react';
import { GanttParams, GanttParamsFields, VisOptionsProps } from '../gantt_vis_type';
import { Field } from 'src/plugins/dashboard/public/types';

export function GanttChartEditor({ aggs, stateParams, setValue }: VisOptionsProps<GanttParams>) {
  const fieldOptions = aggs.indexPattern.fields.map((field: Field) => {
    return {
      value: field.name,
      inputDisplay: field.name,
    };
  });

  return (
    <>
      <EuiPanel paddingSize="s">
        <EuiTitle size="xs">
          <h3>Metrics</h3>
        </EuiTitle>
        <EuiSpacer size="s" />
        <EuiFormRow label="Event" helpText="Gantt chart allows you to compare schedules of the selected field.">
          <EuiSuperSelect
            options={fieldOptions}
            valueOfSelected={stateParams.labelField || 'select'}
            onChange={(value) => setValue('labelField', value)}
          />
        </EuiFormRow>
        <EuiFormRow label="Start time" helpText="Select a timestamp field to represent the beginning of a schedule.">
          <EuiSuperSelect
            options={fieldOptions}
            valueOfSelected={stateParams.startTimeField}
            onChange={(value) => setValue('startTimeField', value)}
          />
        </EuiFormRow>
        <EuiFormRow label="Duration" helpText="Value of duration field must be a time interval that can be added to the start timestamp field.">
          <EuiSuperSelect
            options={fieldOptions}
            valueOfSelected={stateParams.durationField}
            onChange={(value) => setValue('durationField', value)}
          />
        </EuiFormRow>
      </EuiPanel>

      <EuiSpacer size="s" />

      <EuiPanel paddingSize="s">
        <EuiTitle size="xs">
          <h3>Results</h3>
        </EuiTitle>
        <EuiSpacer size="s" />
        <EuiFormRow label="View number of results" helpText="Results ordered by descending start time.">
          <EuiFieldNumber
            value={stateParams.size}
            onChange={(e) => setValue('size', parseInt(e.target.value, 10))}
          />
        </EuiFormRow>
      </EuiPanel>
    </>
  );
}
