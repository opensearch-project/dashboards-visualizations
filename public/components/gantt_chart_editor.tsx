/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiCompressedFieldNumber,
  EuiCompressedFormRow,
  EuiPanel,
  EuiSpacer,
  EuiCompressedSuperSelect,
  EuiTitle,
} from '@elastic/eui';
import React from 'react';
import { Field } from 'src/plugins/dashboard/public/types';
import { GanttParams, VisOptionsProps } from '../gantt_vis_type';

export function GanttChartEditor({ aggs, stateParams, setValue }: VisOptionsProps<GanttParams>) {
  const fieldOptions = aggs.indexPattern.fields.map((field: Field) => {
    return {
      value: field.name,
      inputDisplay: field.name,
    };
  });

  return (
    <>
      <EuiPanel paddingSize="s">
        <EuiTitle size="xs">
          <h3>Metrics</h3>
        </EuiTitle>
        <EuiSpacer size="s" />
        <EuiCompressedFormRow
          label="Event"
          helpText="Gantt chart allows you to compare schedules of the selected field."
        >
          <EuiCompressedSuperSelect
            data-test-subj="gantt-chart-editor-label-field"
            options={fieldOptions}
            valueOfSelected={stateParams.labelField || 'select'}
            onChange={(value) => setValue('labelField', value)}
          />
        </EuiCompressedFormRow>
        <EuiCompressedFormRow
          label="Start time"
          helpText="Select a timestamp field to represent the beginning of a schedule."
        >
          <EuiCompressedSuperSelect
            data-test-subj="gantt-chart-editor-start-time-field"
            options={fieldOptions}
            valueOfSelected={stateParams.startTimeField}
            onChange={(value) => setValue('startTimeField', value)}
          />
        </EuiCompressedFormRow>
        <EuiCompressedFormRow
          label="Duration"
          helpText="Value of duration field must be a time interval that can be added to the start timestamp field."
        >
          <EuiCompressedSuperSelect
            data-test-subj="gantt-chart-editor-duration-field"
            options={fieldOptions}
            valueOfSelected={stateParams.durationField}
            onChange={(value) => setValue('durationField', value)}
          />
        </EuiCompressedFormRow>
      </EuiPanel>

      <EuiSpacer size="s" />

      <EuiPanel paddingSize="s">
        <EuiTitle size="xs">
          <h3>Results</h3>
        </EuiTitle>
        <EuiSpacer size="s" />
        <EuiCompressedFormRow
          label="View number of results"
          helpText="Results ordered by descending start time."
        >
          <EuiCompressedFieldNumber
            data-test-subj="gantt-chart-editor-size-field"
            value={stateParams.size}
            onChange={(e) => setValue('size', parseInt(e.target.value, 10))}
          />
        </EuiCompressedFormRow>
      </EuiPanel>
    </>
  );
}
