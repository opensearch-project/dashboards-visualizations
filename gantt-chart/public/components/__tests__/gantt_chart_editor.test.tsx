/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { fireEvent, render } from '@testing-library/react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { mockEditorProps } from '../../../test/mocks/mockData';
import { GanttChartEditor } from '../gantt_chart_editor';

describe('<GanttChartEditor /> spec', () => {
  configure({ adapter: new Adapter() });

  it('renders the component', () => {
    const setValue = jest.fn();
    const utils = render(<GanttChartEditor {...mockEditorProps(setValue)} />);
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('changes label field', () => {
    const setValue = jest.fn();
    const props = mockEditorProps(setValue);
    props.stateParams.labelField = '';
    const utils = render(<GanttChartEditor {...props} />);
    utils.getByTestId('gantt-chart-editor-label-field').click();
    utils.getByText('test-field-1').click();
    expect(setValue).toBeCalledWith('labelField', 'test-field-1');
  });

  it('changes start time field', () => {
    const setValue = jest.fn();
    const utils = render(<GanttChartEditor {...mockEditorProps(setValue)} />);
    utils.getByTestId('gantt-chart-editor-start-time-field').click();
    utils.getByText('test-field-2').click();
    expect(setValue).toBeCalledWith('startTimeField', 'test-field-2');
  });

  it('changes duration field', () => {
    const setValue = jest.fn();
    const utils = render(<GanttChartEditor {...mockEditorProps(setValue)} />);
    utils.getByTestId('gantt-chart-editor-duration-field').click();
    utils.getByText('test-field-3').click();
    expect(setValue).toBeCalledWith('durationField', 'test-field-3');
  });

  it('changes size field', () => {
    const setValue = jest.fn();
    const utils = render(<GanttChartEditor {...mockEditorProps(setValue)} />);
    const field = utils.getByTestId('gantt-chart-editor-size-field');
    fireEvent.change(field, { target: { value: 3 } });
    expect(setValue).toBeCalledWith('size', 3);
  });
});
