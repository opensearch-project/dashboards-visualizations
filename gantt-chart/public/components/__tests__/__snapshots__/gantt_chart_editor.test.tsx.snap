// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GanttChartEditor /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--paddingSmall euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
>
  <h3
    class="euiTitle euiTitle--xsmall"
  >
    Metrics
  </h3>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        Event
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
      >
        <div
          class="euiPopover__anchor"
        >
          <div>
            <input
              id="random_html_id"
              type="hidden"
              value="spanID"
            />
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <span
                  class="euiScreenReaderOnly"
                  id="random_html_id"
                >
                  Select an option: , is selected
                </span>
                <button
                  aria-describedby="random_html_id-help-0"
                  aria-haspopup="true"
                  aria-labelledby="random_html_id random_html_id"
                  class="euiSuperSelectControl"
                  data-test-subj="gantt-chart-editor-label-field"
                  type="button"
                />
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <span
                    class="euiFormControlLayoutCustomIcon"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFormHelpText euiFormRow__text"
        id="random_html_id-help-0"
      >
        Gantt chart allows you to compare schedules of the selected field.
      </div>
    </div>
  </div>
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        Start time
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
      >
        <div
          class="euiPopover__anchor"
        >
          <div>
            <input
              id="random_html_id"
              type="hidden"
              value="startTime"
            />
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <span
                  class="euiScreenReaderOnly"
                  id="random_html_id"
                >
                  Select an option: , is selected
                </span>
                <button
                  aria-describedby="random_html_id-help-0"
                  aria-haspopup="true"
                  aria-labelledby="random_html_id random_html_id"
                  class="euiSuperSelectControl"
                  data-test-subj="gantt-chart-editor-start-time-field"
                  type="button"
                />
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <span
                    class="euiFormControlLayoutCustomIcon"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFormHelpText euiFormRow__text"
        id="random_html_id-help-0"
      >
        Select a timestamp field to represent the beginning of a schedule.
      </div>
    </div>
  </div>
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        Duration
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
      >
        <div
          class="euiPopover__anchor"
        >
          <div>
            <input
              id="random_html_id"
              type="hidden"
              value="duration"
            />
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <span
                  class="euiScreenReaderOnly"
                  id="random_html_id"
                >
                  Select an option: , is selected
                </span>
                <button
                  aria-describedby="random_html_id-help-0"
                  aria-haspopup="true"
                  aria-labelledby="random_html_id random_html_id"
                  class="euiSuperSelectControl"
                  data-test-subj="gantt-chart-editor-duration-field"
                  type="button"
                />
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <span
                    class="euiFormControlLayoutCustomIcon"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFormHelpText euiFormRow__text"
        id="random_html_id-help-0"
      >
        Value of duration field must be a time interval that can be added to the start timestamp field.
      </div>
    </div>
  </div>
</div>
`;
